#198/1 - Add new model classes to support REST API updates without breaking changes

hamorillo opened the pull request:
Closes #198 

### Description
This is the first PR of a set of three (reviewing them in order will help), whose aim is to refactor all the necessary stuff to allow us to update the REST API without breaking changes or releasing a new major version. This conversation started in #184.

The idea is to create model classes that allow us to add new fields with binary or source compatibility. This PR is highly inspired by [this post](https://jakewharton.com/public-api-challenges-in-kotlin/).

As we were already using OpenApi templates, this PR adds and refactors templates to generate model classes that can easily updated.

This PR only updates templates and generates classes that are not used yet. That will be the next PR.

### Testing Steps
- Code review and CI should be enough.

hamorillo committed:
Add "data class" OpenApi templates to modify them

hamorillo committed:
Update OpenApi templates to generate non-breaking models

Using OpenApi generator templates, we can modify how the models are generated to create them without making a new major release when we need to add new fields.

Great documentation can be found in:
https://jakewharton.com/public-api-challenges-in-kotlin/

hamorillo commented on gravatar/build.gradle.kts:
```diff
@@ -119,7 +119,7 @@ openApiGenerate {
             "serializationLibrary" to "gson",
             "groupId" to "com.gravatar",
-            "packageName" to "com.gravatar.api",
+            "packageName" to "com.gravatar.restapi",
```
What do you think about `restapi` as the name for the new package?

AdamGrzybkowski replied:
I like it 👍

AdamGrzybkowski commented on gravatar/openapi/templates/data_class.mustache:
```diff
@@ -144,6 +145,34 @@ import {{packageName}}.infrastructure.ITransformForStorage
-{{#vendorExtensions.x-has-data-class-body}}
+
+{{^isEnum}}
+	override fun toString() : String = "{{#allVars}}{{className}}{{name}}=${{name}}{{^-last}}, {{/-last}}{{/allVars}}"
```
I think this needs to be modified. The generated value of the `toString` function doesn't have the class name. The end result is "param1=value1" instead of "ClassName(param1=value1)"

hamorillo replied:
Good catch! Fixed!

AdamGrzybkowski commented on gravatar/openapi/templates/data_class.mustache:
```diff
@@ -60,14 +61,14 @@ import {{packageName}}.infrastructure.ITransformForStorage
 {{/isDeprecated}}
 {{>additionalModelTypeAnnotations}}
-{{#nonPublicApi}}internal {{/nonPublicApi}}{{#discriminator}}interface{{/discriminator}}{{^discriminator}}public data class{{/discriminator}} {{classname}}{{^discriminator}} (
+{{#nonPublicApi}}internal {{/nonPublicApi}}{{#discriminator}}interface{{/discriminator}}{{^discriminator}}public class{{/discriminator}} {{classname}}{{^discriminator}} private constructor(
```
Curious but the `public` keyword has no condition so if we use `nonPublicApi` we will end up with `internal public class`, right?

hamorillo replied:
Good point. 

I've added the condition, but to be honest, I am not 100% sure that the kotlin generator supports it. I don't see it on the [kotlin docs](https://openapi-generator.tech/docs/generators/kotlin), and I couldn't use it🤷🏽‍♂️. Anyway, it looks better with the condition, just in case.

AdamGrzybkowski reviewed (approved):
Great work with those templates! I left some comments though nothing critical.

hamorillo committed:
Fix "toString" method in OpenApi templates

The generated value of the toString function didn't have the class name. The end result was "param1=value1" instead of "ClassName(param1=value1)"

hamorillo committed:
Fix OpenApi template to only add the "public" when "nonPublicApi"
